### PR TITLE
feat: add GCS sink

### PR DESCRIPTION
**Pull request by Chief-Rishab**

### CODE REVIEW

File: plugins/sinks/gcs/client.go [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+import (
+	"context"
+
+	"cloud.google.com/go/storage"
+	"github.com/pkg/errors"
+	"google.golang.org/api/option"
+)
+
+type GCSClient interface {
```
Comment (kushsharma) [inline plugins/sinks/gcs/client.go:11]: We can call this `Writer`?

File: plugins/sinks/gcs/README.md [hunk 2]
```diff
@@ -0,0 +1,41 @@
+# GCS
+
+Sinks json data to a file in `ndjson` format in a Google Cloud Storage bucket
```
Comment (kushsharma) [inline plugins/sinks/gcs/README.md:3]: Maybe we can rephrase this as:

Sinks json data to a file as [ndjson](http://ndjson.org/) format in Google Cloud Storage bucket

File: plugins/sinks/gcs/client.go [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+func newGCSClient(ctx context.Context, serviceAccountJSON []byte, bucketname string, filepath string) (GCSClient, error) {
+	client, err := storage.NewClient(ctx, option.WithCredentialsJSON(serviceAccountJSON))
+	if err != nil {
+		return nil, errors.Wrap(err, "error in creating client")
+	}
+
+	writer := client.Bucket(bucketname).Object(filepath).NewWriter(ctx)
+
+	return &gcsClient{
+		client: client,
```
Comment (kushsharma) [inline plugins/sinks/gcs/client.go:30]: I don't see client getting used anywhere, should we get rid of it?

File: plugins/sinks/gcs/client.go [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+	WriteData([]byte) error
+	Close() error
+}
+
+type gcsClient struct {
+	client *storage.Client
+	writer *storage.Writer
+}
+
+func newGCSClient(ctx context.Context, serviceAccountJSON []byte, bucketname string, filepath string) (GCSClient, error) {
```
Comment (kushsharma) [inline plugins/sinks/gcs/client.go:21]: You should return `*gcsClient` instead of interface.
> Remember always return struct and accept interfaces in args

File: plugins/sinks/gcs/client.go [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+	return &gcsClient{
+		client: client,
+		writer: writer,
+	}, nil
+}
+
+func (c *gcsClient) WriteData(jsonBytes []byte) error {
+	if _, err := c.writer.Write(jsonBytes); err != nil {
+		return errors.Wrap(err, "error in writing json data to an object")
```
Comment (kushsharma) [inline plugins/sinks/gcs/client.go:37]: the client should not know if the data is json or something else, its just raw bytes for it.

File: plugins/sinks/gcs/client.go [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+	return nil
+}
+
+func (c *gcsClient) Close() error {
+	if err := c.writer.Close(); err != nil {
+		return errors.Wrap(err, "error closing the writer")
+	}
+
+	return nil
+}
```
Comment (kushsharma) [inline plugins/sinks/gcs/client.go:49]: Maybe simplify this as
```
func (c *gcsClient) Close() error {
	return  c.writer.Close()
}
```

File: plugins/sinks/gcs/client.go (hunk 3 above)
Comment (Chief-Rishab) [reply plugins/sinks/gcs/client.go:30]: yes, its's only used to create the writer, will update the struct

File: plugins/sinks/gcs/gcs.go [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+		}
+		s.config.ServiceAccountJSON = string(serviceAccountJSON)
+	}
+	return nil
+}
+
+func (s *Sink) resolveBucketandObjectNames() (string, string) {
+	dirs := strings.Split(s.config.Path, "/")
+	bucketname := dirs[0]
+	timestamp := time.Now().Format("2006.01.02 15:04:05")
```
Comment (kushsharma) [inline plugins/sinks/gcs/gcs.go:107]: I would suggest using a pre-defined time RFC format, specially the one without space.

File: plugins/sinks/gcs/gcs.go [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+
+	if s.config.ObjectPrefix != "" {
+		s.config.ObjectPrefix = s.config.ObjectPrefix + "-"
+	}
```
Comment (kushsharma) [inline plugins/sinks/gcs/gcs.go:111]: What if user has provided prefix as `hello-`, you will add double hyphens?

File: plugins/sinks/gcs/gcs.go [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+
+	objectname := s.config.ObjectPrefix + timestamp + ".ndjson"
```
Comment (kushsharma) [inline plugins/sinks/gcs/gcs.go:113]: use `fmt.Sprintf` to build strings.

File: plugins/sinks/gcs/gcs.go [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+	if len(dirs) > 1 {
+		objectname = dirs[len(dirs)-1] + "/" + s.config.ObjectPrefix + timestamp + ".ndjson"
```
Comment (kushsharma) [inline plugins/sinks/gcs/gcs.go:115]: use `fmt.Sprintf` to build strings.

File: plugins/sinks/gcs/gcs.go [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+
+	if s.client, err = newGCSClient(ctx, []byte(s.config.ServiceAccountJSON), bucketname, objectname); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (s *Sink) validateServiceAccountKey() error {
+
```
Comment (kushsharma) [inline plugins/sinks/gcs/gcs.go:89]: unnecessary space?

File: plugins/sinks/gcs/gcs.go (hunk 8 above)
Comment (Chief-Rishab) [reply plugins/sinks/gcs/gcs.go:111]: It should be a single hyphen only. I will add a check for that

File: plugins/sinks/gcs/gcs.go [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+	}
+	return nil
+}
+
+func (s *Sink) Close() (err error) {
+	if err := s.client.Close(); err != nil {
+		return err
+	}
+	return nil
+}
```
Comment (kushsharma) [inline plugins/sinks/gcs/gcs.go:153]:
```go
return s.client.Close()
```

File: plugins/sinks/gcs/gcs.go (hunk 7 above)
Comment (Chief-Rishab) [reply plugins/sinks/gcs/gcs.go:107]: RFC3339? Output format is "2006-01-02T15:04:05Z07:00" ?

File: plugins/sinks/gcs/gcs.go [hunk 13]
```diff
@@ -0,0 +1,161 @@
...
+		serviceAccountJSON, err := base64.StdEncoding.DecodeString(s.config.ServiceAccountBase64)
+		if err != nil || len(serviceAccountJSON) == 0 {
+			return errors.Wrap(err, "failed to decode base64 service account")
+		}
+		s.config.ServiceAccountJSON = string(serviceAccountJSON)
+	}
+	return nil
+}
+
+func (s *Sink) resolveBucketandObjectNames() (string, string) {
```
Comment (kushsharma) [inline plugins/sinks/gcs/gcs.go:104]: A better name could be `resolveBucketPath`

File: plugins/sinks/gcs/gcs.go [hunk 14]
```diff
@@ -0,0 +1,161 @@
...
+	dirs := strings.Split(s.config.Path, "/")
+	bucketname := dirs[0]
```
Comment (kushsharma) [inline plugins/sinks/gcs/gcs.go:106]: A notation pretty common in object storage is `gcs://bucketname` or `s3://bucketname` I guess this line would fail here. Can we use `url.Parse(...)`?

File: plugins/sinks/gcs/gcs.go [hunk 15]
```diff
@@ -0,0 +1,161 @@
...
+	objectprefix := s.config.ObjectPrefix 
+	
+	if objectprefix != "" && objectprefix[len(objectprefix)-1] !='-'{
+		objectprefix = fmt.Sprintf("%s-", objectprefix)
+	}
+	
+	objectname := fmt.Sprintf("%s%s.ndjson", objectprefix, timestamp)
+	
+	if len(dirs) > 1 {
+		objectname = fmt.Sprintf("%s/%s%s.ndjson", dirs[len(dirs)-1], objectprefix, timestamp)
```
Comment (kushsharma) [inline plugins/sinks/gcs/gcs.go:118]: How will `dirs[len(dirs)-1]` work for a path like `gcs://bucketname/path1/path2/`

File: plugins/sinks/gcs/gcs.go [hunk 16]
```diff
@@ -0,0 +1,161 @@
...
+		}
+
+		if err := s.writer.WriteData([]byte("\n")); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (s *Sink) Close() (err error) {
```
Comment (kushsharma) [inline plugins/sinks/gcs/gcs.go:151]: `(err error)` could be just `error`

**Review by kushsharma — CHANGES_REQUESTED** [review]

File: plugins/sinks/gcs/gcs.go (hunk 15 above)
Comment (Chief-Rishab) [reply plugins/sinks/gcs/gcs.go:118]: The default path format I assumed was the one I got from GCS bucket/folder copy path options. If that's the required case, will handle the gcs prefix

**Comment by kushsharma** [comment]
LGTM. Nice work.
